File: packages/core/src/projection.py
"""Reconstruct the file content a tool call would produce, without writing it.

The gate judges the *projected* content, never the raw edit instruction: Write
supplies the whole file; Edit is replayed against the current on-disk file. When
the outcome is undecidable — an Edit whose ``old_string`` is not in the file, or
a missing file — the gate declines to judge. The harness will reject that edit
on its own, and we never double-judge.
"""
from dataclasses import dataclass
from pathlib import Path

WRITE = "Write"
EDIT = "Edit"


@dataclass(frozen=True)
class Projection:
    """The before/after content of a decidable, gated tool call."""

    path: str
    tool_name: str
    current: str    # on-disk content before the call ("" for a new file)
    projected: str  # content the call would produce


def project(tool_name, tool_input):
    """Return a :class:`Projection`, or ``None`` when the call is not ours or
    is undecidable."""
    path = tool_input.get("file_path")
    if not path:
        return None

    if tool_name == WRITE:
        content = tool_input.get("content")
        if content is None:
            return None
        current = _read(path)
        return Projection(path, WRITE, current or "", content)

    if tool_name == EDIT:
        old_string = tool_input.get("old_string")
        if not old_string:
            return None  # empty/missing anchor -> undecidable
        current = _read(path)
        if current is None or old_string not in current:
            return None  # missing file, or a stale anchor -> undecidable
        new_string = tool_input.get("new_string", "")
        if tool_input.get("replace_all", False):
            projected = current.replace(old_string, new_string)
        else:
            projected = current.replace(old_string, new_string, 1)
        return Projection(path, EDIT, current, projected)

    return None


def _read(path):
    try:
        return Path(path).read_text(encoding="utf-8")
    except (FileNotFoundError, IsADirectoryError):
        return None

File: packages/core/src/projection.ts
/**
 * Reconstruct the file content a tool call would produce, without writing it.
 *
 * The gate judges the *projected* content, never the raw edit instruction:
 * Write supplies the whole file; Edit is replayed against the current on-disk
 * file. When the outcome is undecidable — an Edit whose `old_string` is not
 * in the file, or a missing file — the gate declines to judge. The harness
 * will reject that edit on its own, and we never double-judge.
 */

export const WRITE = "Write";
export const EDIT = "Edit";

export interface ToolInput {
  file_path?: string;
  content?: string;
  old_string?: string;
  new_string?: string;
  replace_all?: boolean;
}

/** The before/after content of a decidable, gated tool call. */
export interface Projection {
  path: string;
  toolName: string;
  /** On-disk content before the call ("" for a new file). */
  current: string;
  /** Content the call would produce. */
  projected: string;
}

/** Return a Projection, or null when the call is not ours or is undecidable. */
export async function project(toolName: string, toolInput: ToolInput): Promise<Projection | null> {
  const path = toolInput.file_path;
  if (!path) return null;

  if (toolName === WRITE) {
    const { content } = toolInput;
    if (content === undefined || content === null) return null;
    const current = await read(path);
    return { path, toolName: WRITE, current: current ?? "", projected: content };
  }

  if (toolName === EDIT) {
    const oldString = toolInput.old_string;
    if (!oldString) return null; // empty/missing anchor -> undecidable
    const current = await read(path);
    if (current === null || !current.includes(oldString)) {
      return null; // missing file, or a stale anchor -> undecidable
    }
    const newString = toolInput.new_string ?? "";
    const projected = toolInput.replace_all
      ? current.replaceAll(oldString, newString)
      : current.replace(oldString, newString);
    return { path, toolName: EDIT, current, projected };
  }

  return null;
}

async function read(path: string): Promise<string | null> {
  try {
    const file = Bun.file(path);
    if (!(await file.exists())) return null; // missing file or a directory
    return await file.text();
  } catch {
    return null;
  }
}
